PLAT-1214: Support Azure Container Registry

fraenkel opened the pull request:
ACR is a supported container registry.
    The credentials are currently pulled from the environment and used to
    obtain a refresh token for login.

Commit:
PLAT-1214: ACR e2e tests

Azure ACR push/pull tests

Commit:
PLAT-1214: Support Azure Container Registry

ACR is a supported container registry.
The credentials are currently pulled from the environment and used to
obtain a refresh token for login.

Retry obtaining the initial credential when using MSI. There seems to be
some delay at times to gain access to 169.254.169.254.

fraenkel commented:
This is now ready to go.
Verified that with repocloner enabled, we can push and pull from ACR.

steved commented on internal/cloud/acr/acr.go:
```diff
@@ -0,0 +1,164 @@
+	}
+
+	// verify headers
+	if len(authParams["service"]) == 0 {
```
just curious, why `len` vs. `== ""`?

fraenkel replied:
this code was lifted. I will make it more go-ish.

fraenkel commented:
This is can't be merged until #86 is merged.

Commit:
PLAT-2546: Update aws modules

Commit:
PLAT-2546: AWS Region can come from IMDS

Enable the retrieval of the current AWS region from IMDS.

Configure a default AWS_REGION for CI to run properly.
169.254.169.254 (IMDS) is available but returns a 404 for the Region.

Commit:
Fix racy amap publish test

There was no guarantee that the error appeared before Push was checking
for an error.
Place the error in the buffered channel prior to Push.

sonnysideup reviewed (commented):
I think the only big questions I have are related to (a) adding a `panic` and (b) configuring ACR during build time.

Other than that, it seems like the cloud registry bits were refactored so that registration is initiated by the controller at start time instead of at runtime during invocation. I think the previous design was nice because it didn't require _initiation of all of the providers_ but the code was a bit onerous in its complexity and suffered from some other design flaws. This newer approach seems fine to me.

fraenkel commented on .circleci/config.yml:
```diff
@@ -24,6 +26,31 @@ jobs:
     steps:
       - checkout
+      - azure-cli/install
+      - run:
+          name: Create ACR
+          command: |
+            az login --service-principal -u "${AZURE_CLIENT_ID}" -p "${AZURE_CLIENT_SECRET}" --tenant "${AZURE_TENANT_ID}"
+
+            export ACR_VARS=/tmp/workspace/acr_vars
+            mkdir -p $(dirname $ACR_VARS)
+            echo "export ACR_NAME=forge${CIRCLE_BUILD_NUM}" >> $ACR_VARS
+            echo "export ACR_RG=forge${CIRCLE_BUILD_NUM}" >> $ACR_VARS
+            echo "export ACR_LOCATION=westus2" >> $ACR_VARS
+            . $ACR_VARS
+            az group create -n ${ACR_RG} -l ${ACR_LOCATION}
+            az acr create -n ${ACR_NAME} -g ${ACR_RG} -l ${ACR_LOCATION} --sku Basic
+            echo "export ACR_REGISTRY=$(az acr show -n ${ACR_NAME} -g ${ACR_RG} --query loginServer)" >> $ACR_VARS
+            cat $ACR_VARS >> $BASH_ENV
```
For units, its automatically skipped if you dont have Azure creds.
For CI, we do so I setup ACR for both e2e and units.

sonnysideup replied:
> For units, its automatically skipped if you dont have Azure creds.

Does that mean our unit tests "hit" ACR when Azure credentials are available and configured?

fraenkel replied:
yes.

fraenkel commented on internal/cloud/acr/acr.go:
```diff
@@ -0,0 +1,164 @@
+		}
+		if err != nil {
+			// IMDS can take some time to setup, restart the process
+			panic(err)
```
Why should the error matter? 
And this begins the issue. 
In the case you provide a full set of credentials, we don't validate them. This will appear as a failure when attempting to process a CIB unless we attempt to obtain a token during registration which might be a good idea for consistency.
When using MSI, we can fail to obtain the credentials because the information is invalid or the MSI endpoint is not setup yet. For either, we should exit given you intend to use MSI and its failing.

sonnysideup replied:
>  When using MSI, we can fail to obtain the credentials because the information is invalid or the MSI endpoint is not setup yet. For either, we should exit given you intend to use MSI and its failing.

The name of the error doesn't really matter to me as much as the message. I would assume that, as a consumer of this API, if the MSI endpoint is not setup, then tell me so I can figure out what to do about it. It just so happens we want to exist but I think the controller clearly exiting because MSI is not ready is more obvious than a `panic` causing the process to crash exit. 

Does that make sense?

fraenkel replied:
Welcome to Azure. The issue isn't whether its setup or not. It's a timing issue with Azure. MSI can be expected but take seconds after the instance is up before it's connected.  The docs for MSI even state that you need to retry if you get a Connection Refused which is how it appears.
I have a slightly different philosophy, if you have configured something and it fails, do not pass go.
If you provide details for MSI and it fails for whatever reason, error out. Let the user decide to either remove the configuration or fix whatever is wrong. Doing anything else means we start without ACR when you want it and it will surface as errors in Domino which aren't obvious.

fraenkel commented on internal/cloud/registry.go:
```diff
@@ -0,0 +1,36 @@
+	"github.com/docker/docker/api/types"
+)
+
+var errNoLoader = errors.New("no loader found")
```
Until we actually need it, there isn't any reason to expose it.

sonnysideup reviewed (approved):
Looks good. I would just fix the test suite and then you're go to merge
